### PR TITLE
SpringBoot

### PR DESCRIPTION
**Pull request by dereksong88**
https://www.notion.so/7da38047a5c94fbaa0cc358cfe06afc1?v=fa1fa99ce90c45bf885063963d54586d

### CODE REVIEW

File: springboot/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+
```
Comment (jarviscanada) [inline springboot/Dockerfile:1]: `git rm  springboot/target` 
do not commit genereated files

File: springboot/README.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
++ Make sure docker is running by the following command:\
+`systemctl status docker || systemctl start docker`
++ Create a docker network which allows docker containers to communicate with each other by the command:\
+`docker network create --driver bridge trading-net`\
+Verify\
+`docker network ls`
+
++ Build docker image `trading-psql`. It is for PostgreSQL database which contains all market data and user information.\
+`cd ./springboot/psql
```
Comment (jarviscanada) [inline springboot/README.md]:
```
google search and use markdown code block syntax.
fix this for all occurrences 
```

File: springboot/src/main/java/ca/jrvs/apps/trading/AppConfig.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        cm.setMaxTotal(50);
+        cm.setDefaultMaxPerRoute(50);
+        return cm;
+       }
+
+//    public static void main(String[] args) {
+//        System.out.println(System.getenv("PSQL_URL"));
+//        System.out.println(System.getenv("PSQL_USER"));
+//        System.out.println(System.getenv("PSQL_PASSWORD"));
+//    }
```
Comment (jarviscanada) [inline springboot/src/main/java/ca/jrvs/apps/trading/AppConfig.java]: remove all unused comments
fix for all occurrences

File: springboot/src/main/java/ca/jrvs/apps/trading/Application.java [hunk 4]
```diff
@@ -1,10 +1,38 @@
...
+import org.springframework.boot.CommandLineRunner;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration;
+import org.springframework.boot.autoconfigure.jdbc.JdbcTemplateAutoConfiguration;
+import org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaAutoConfiguration;
+import org.springframework.stereotype.Component;
 
-  // Your program begins with a call to main().
-  // Prints "Hello, World" to the terminal window.
```
Comment (jarviscanada) [inline springboot/src/main/java/ca/jrvs/apps/trading/Application.java:6]: clean up

File: springboot/src/main/java/ca/jrvs/apps/trading/dao/MarketDataDao.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+@Repository
+public class MarketDataDao implements CrudRepository<IexQuote, String> {
+    private static final String IEX_BATCH_PATH = "/stock/market/batch?symbols=%s&types=quote&token=";
+    private static String IEX_BATCH_URL;
+
+    private Logger logger = LoggerFactory.getLogger(MarketDataDao.class);
+    private HttpClientConnectionManager httpClientConnectionManager;
+    //private PoolingHttpClientConnectionManager httpClientConnectionManager;
+
+    //private String[] tickerFields = {"GOOGL", "FB", "AAPL", "MMM", "AMD", "APA"};
```
Comment (jarviscanada) [inline springboot/src/main/java/ca/jrvs/apps/trading/dao/MarketDataDao.java]: clean up

File: springboot/src/main/java/ca/jrvs/apps/trading/dao/MarketDataDao.java [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        if (response.getEntity() == null) {
+            throw new RuntimeException("Empty response body");
+        }
+
+        try {
+            jsonStr = Optional.of(EntityUtils.toString(response.getEntity()));
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to convert entity to String", e);
+        }
+        return jsonStr;
```
Comment (jarviscanada) [inline springboot/src/main/java/ca/jrvs/apps/trading/dao/MarketDataDao.java]: throwing runtime exception is a bad practice.
use DataRetrievalFailureException instead

Comment (jarviscanada) [reply springboot/src/main/java/ca/jrvs/apps/trading/dao/MarketDataDao.java]:
```
  /**
   * Execute a get and return http entity/body as a string
   *
   * Tip: use EntityUtils.toString to process HTTP entity
   *
   * @param url resource URL
   * @return http response boy or Optional.empty for 404 response
   * @throws DataRetrievalFailureException if HTTP failed or status code is unexpected
   */
  private Optional<String> executeHttpGet(String url) {
    try (CloseableHttpClient httpClient = getHttpClient()) {
      HttpGet httpGet = new HttpGet(url);
      try (CloseableHttpResponse response = httpClient.execute(httpGet)) {
        switch (response.getStatusLine().getStatusCode()) {
          case 200:
            //EntityUtils toString will also close inputStream in Entity
            String body = EntityUtils.toString(response.getEntity());
            return Optional.ofNullable(body);
          case 404:
            return Optional.empty();
          default:
            throw new DataRetrievalFailureException(
                "Unexpected status:" + response.getStatusLine().getStatusCode());
        }
      }
    } catch (IOException e) {
      throw new DataRetrievalFailureException("Unable Http execution error", e);
    }
  }
```

File: springboot/src/main/java/ca/jrvs/apps/trading/dao/MarketDataDao.java [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        }
+        //construct JSONObject from JSON String
+        JSONObject jsonObject = new JSONObject(str.get());
+        for (String s : tickers) {
+            singeJsonStr = jsonObject.getJSONObject(s).getJSONObject("quote").toString();
+            try {
+                iexQuote = toObjectFromJson(singeJsonStr, IexQuote.class);
+
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (jarviscanada) [inline springboot/src/main/java/ca/jrvs/apps/trading/dao/MarketDataDao.java]: do not print. use logger instead. please review java coding style from Notion

File: springboot/src/main/java/ca/jrvs/apps/trading/dao/QuoteDao.java [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    @Autowired
+    public QuoteDao(DataSource dataSource) {
+        jdbcTemplate = new JdbcTemplate(dataSource);
+        simpleJdbcInsert = new SimpleJdbcInsert(dataSource).withTableName(TABLE_NAME);
+    }
+
+    public static void main(String[] args) {
+//        String url = "jdbc:postgresql://localhost:5432/jrvstrading";
+//        String user = "postgres";
+//        String password = "password";
```
Comment (jarviscanada) [inline springboot/src/main/java/ca/jrvs/apps/trading/dao/QuoteDao.java]: clean up

File: springboot/src/main/java/ca/jrvs/apps/trading/AppConfig.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
+package ca.jrvs.apps.trading;
+
+import ca.jrvs.apps.trading.dao.MarketDataDao;
+import ca.jrvs.apps.trading.model.config.MarketDataConfig;
+import ca.jrvs.apps.trading.service.QuoteService;
```
Comment (jarviscanada) [inline springboot/src/main/java/ca/jrvs/apps/trading/AppConfig.java]: run a code auto format from you IDE
https://www.notion.so/jarviscanada/IntelliJ-IDEA-Guide-2e124ee013ee4ef3b8e1ee55e3b494aa#d77de86780894612a7d72c479ff67d26

**Review by jarviscanada — APPROVED** [review]
this PR is too big( too many file changes)

File: springboot/src/main/java/ca/jrvs/apps/trading/AppConfig.java (hunk 9 above)
Comment (dereksong88) [reply springboot/src/main/java/ca/jrvs/apps/trading/AppConfig.java]: @jarviscanada Hello Edward, thank you so much for pointing out all the nonconformances. I have corrected them.

**Pull request merged**